### DISCUSSION TITLE
RFC: native algebra data type/tagged union/sum type/enum support

### DISCUSSION THREAD

**Discussion by Roger-luo** [JuliaLang/julia#48883] [RFC: features for discussion]
I tried to find an issue that contains a discussion for native support of this but didn't find any. 

I think this is a very well-explored feature in the community already and people seem to have been using it for many different cases through packages, I collect some implementations by searching JuliaHub (that might not be complete, so please feel free to add yours):

- `Expronicon.ADT` (my own): [discourse post](https://discourse.julialang.org/t/ann-expronicon-v0-8-type-stable-rust-like-algebra-data-type-pretty-printing-julia-expr-more/92647)
- `Unityper` used by Symbolics.jl: https://github.com/YingboMa/Unityper.jl
- `SumTypes` (by @MasonProtter ): https://github.com/MasonProtter/SumTypes.jl
- `MLStyle` (by @thautwarm ): https://thautwarm.github.io/MLStyle.jl/latest/preview.html#generalized-algebraic-data-types
- [a few packages](https://juliahub.com/ui/Search?q=enum&type=packages) trying to improve the current `enum` should also fall in this category

The reason why I'm requesting opinions on having this supported natively is that I find there are some limitations as a package implementation, mostly because of type-instability created by package implementation and not being able to actually manage the behaviour of allocation, I will refer to this feature as `ADT` in the following context:

I'm not if this is the only solution, but to make it static, the `ADT` requires a special fixed memory layout, similar to a C union, that is in principle the maximum size of the variant it contains. Currently `Expronicon` and `Unityper` implements this as a special structure, which has a size larger than theoretical because there is no access to control the memory layout from Julia (and we probably shouldn't do it...)

On the other hand, supporting generic ADT can be problematic without support from Julia's type inference, taking the `Option<T>` type (similar cases like `Result` etc. are similar) from rust as an example,

```julia
using Expronicon.ADT: @adt

@adt Option{T} begin
    None
    Some(T)
end
```

currently, if we try to make `None` and `Some` share the same Julia struct (which also means the same size), the `None` from `Option{Float64}` will be different from `Option{Int}`, so it seems an unintuitive, I don't have a good solution to this if we allow something like this it will create type instability, but perhaps if this is a native type, then type inference can at least warn people.

Better support of the ADT requires also pattern matching, but I think we can do that separately as long as reflections are implemented for the built-in ADT.

There probably gonna be more design details that the community wants to work on since I'm not the first one proposing this. But I want to open this post so people could push it forward.

**Comment by StefanKarpinski** [comment]
> but perhaps if this is a native type, then type inference can at least warn people.

Type inference in Julia doesn't emit warnings, it's purely for optimization, not type checking.

**Reply by Roger-luo** [reply]
True, I guess this is more for tools like JET then, but I'm also not sure if type inference will help here at all except optimizing memory layout, so just want to throw an idea here and see what people think

**Reply by ToucheSir** [reply]
RE type inference, I think one thing which would help is if we could somehow instruct the compiler to materialize Union fields of parametric types instead of the default of moving the union to the outside. e.g. The current behaviour is:
```julia
julia> struct Wrapper{T} prop::T end

julia> f(x) = Wrapper(x)
f (generic function with 1 method)

julia> code_warntype(f, (Union{String,Int},))
MethodInstance for f(::Union{Int64, String})
  from f(x) in Main at REPL[6]:1
Arguments
  #self#::Core.Const(f)
  x::Union{Int64, String}
Body::Union{Wrapper{Int64}, Wrapper{String}}
1 ─ %1 = Main.Wrapper(x)::Union{Wrapper{Int64}, Wrapper{String}}
└──      return %1
```
Whereas it would be nice to opt into this instead:
```julia
julia> code_warntype(f, (Union{String,Int},))
MethodInstance for f(::Union{Int64, String})
  from f(x) in Main at REPL[6]:1
Arguments
  #self#::Core.Const(f)
  x::Union{Int64, String}
Body::Wrapper{Union{Int64, String}}
1 ─ %1 = Main.Wrapper(x)::Wrapper{Union{Int64, String}}
└──      return %1
```
This would allow us to leverage Julia's existing optimizations for efficient tagged union storage. It wouldn't be a full replacement for dedicated ADTs, but it could obviate a number of cases where people are currently reaching for ADT libraries.

**Reply by MasonProtter** [reply]
Those two types aren't at all equivalent though

**Reply by ToucheSir** [reply]
That's correct, but I don't understand your point? The ask is not to write the exact same code but get different results, but to have some knob somewhere (rewriting `f` is fine) which allows for the second example to work. Currently you can't usually get the compiler to materialize a `Union` unless you know all concrete types involved in it. Functions like `map` cheat by using `_return_type`, but given how sensitive that is to things like `max_methods` I don't feel it's the right tool for the job.

**Reply by MasonProtter** [reply]
Ah I missed the "opt in" part. Yeah, having it be opt-in would be less problematic. However, I suspect there'd be a lot of resistance to even offering that as an optional knob because it would make type inference change the results of computations, and would make it so that changing type inference in future versions could break code which opted into thing thing.

**Reply by ToucheSir** [reply]
In a way we've already reached this crossroads with `_return_type` and how it's used in the Base collections API (`map`, `accumulate`, etc). I feel like there must be a path for making Unions more first-class that isn't so brittle to inference-related changes.

**Reply by MasonProtter** [reply]
Why not just use ADTs?

**Reply by ToucheSir** [reply]
Because they have to be declared statically ahead of time? Say I wanted to replicate differentiable Swift's [representation of stored CFG pullbacks](https://github.com/apple/swift/blob/main/docs/DifferentiableProgrammingImplementation.md#differential-struct-creation). With statically defined ADTs I'd have to resort to very impure `@eval` in generated functions or some such.

Maybe a way forward is to have a `TaggedUnion` type which actually exists at all times?

**Reply by MasonProtter** [reply]
Swift's CFG pullbacks are using ADTs there though. You wouldn't need to `eval` in a generated function. If I understand correctly, the thing
```
enum EnumBB0 {
  case BB1(StructBB1)
  case BB2(StructBB2)
}
```
could be represented in the top level as
```julia
struct Enum{T} where {T}
    tag::Int
    data::T
end
```
and then in running code you can freely write things like
```julia
EnumBB0 = Enum{Union{StructBB1, StructBB2}}
EnumBB1 = Enum{StructBB3}
EnumBB2 = Enum{StructBB4}

BB1(x) = EnumBB0(1, x)
BB2(x) = EnumBB0(2, x)
BB3(x) = EnumBB1(1, x)
BB4(x) = EnumBB2(1, x)
```

**Reply by ToucheSir** [reply]
I think that would still face the same problem because `StructBB1` and `StructBB2` are themselves parametric types? I'm still unclear as to how you'd make an example like the following work (assume return types of `b1` and `b2` are not a priori known):
```julia
julia> struct MyEnum{T}
           tag::Int
           data::T
       end

julia> b1() = (sin, cos)

julia> b2() = (tan,)

julia> function f(x)
           pbs = x > 0 ? b1() : b2()
           tag = x > 0 ? 1 : 2
           return MyEnum(tag, pbs)
       end

julia> @code_warntype f(-1)

MethodInstance for f(::Int64)
  from f(x) in Main at REPL[6]:1
Arguments
  #self#::Core.Const(f)
  x::Int64
Locals
  tag::Int64
  pbs::Union{Tuple{typeof(tan)}, Tuple{typeof(sin), typeof(cos)}}
  @_5::Union{Tuple{typeof(tan)}, Tuple{typeof(sin), typeof(cos)}}
  @_6::Int64
Body::Union{MyEnum{Tuple{typeof(tan)}}, MyEnum{Tuple{typeof(sin), typeof(cos)}}} # union not captured in MyEnum
1 ─       Core.NewvarNode(:(tag))
│         Core.NewvarNode(:(pbs))
│   %3  = (x > 0)::Bool
└──       goto #3 if not %3
2 ─       (@_5 = Main.b1())
└──       goto #4
3 ─       (@_5 = Main.b2())
4 ┄       (pbs = @_5)
│   %9  = (x > 0)::Bool
└──       goto #6 if not %9
5 ─       (@_6 = 1)
└──       goto #7
6 ─       (@_6 = 2)
7 ┄       (tag = @_6)
│   %15 = Main.MyEnum(tag, pbs)::Union{MyEnum{Tuple{typeof(tan)}}, MyEnum{Tuple{typeof(sin), typeof(cos)}}}
└──       return %15
```

**Reply by MasonProtter** [reply]
> I think that would still face the same problem because StructBB1 and StructBB2 are themselves parametric types? 

What would that change? I thought Swift's type system was invariant just like ours.

>  I'm still unclear as to how you'd make an example like the following work (assume return types of b1 and b2 are not a priori known):

I guess you'd do the exact same thing you've requested the compiler do automatically (with an opt-in) and use `return_type`.
```julia
R1 = return_type(b1, Tuple{})
R2 = return_type(b2, Tuple{})
MyEnum{Union{R1, R2}}(tag, pbs)
```
It then becomes your responsibility to write code where the overall result wouldn't be broken by type inference becoming more or less precise, which I think for the use-case you have in mind is not so hard to do.

**Reply by Roger-luo** [reply]
generic ADT is part of the reason why I want some support from the compiler, but not because of the type stability - you cannot have type stability on different type parameters, because they do not have the same memory layout. I think in rust you will have to box them manually whereas in Julia type-instable objects are boxed automatically. 

I think the issue is more on comparison of variants from different type parameters, having proper namespace and proper uninitialized behavior, e.g the `Option{T}` type

```julia
@enum OptionVariantType begin
    None
    Some
end

struct Option{T}
   tag::OptionVariantType
   data::T
end
```

here for the different type parameters, the variant `None` is not considered the same variant because they are from different data layouts. A built-in support of such will force the behavior more consistent and treat `None` from `Option{Float64}` and `None` from `Option{Int}` to be the same.

for the namespace, in `EnumX` this is emulated by generating a Module, but module is not dispatchable, in Expronicon, we instead overload `getproperty(::Type{MyADTType}, name::Symbol)`, which kinda works but breaks a lot of Base functions and because they were expecting to call `getfield(::Type{MyADTType}, name::Symbol)`

for initialization, current the `new` function requires an order of uninitialized data, which is not something we can always have when defining variants, e.g 

```julia
@adt Demo begin
    struct Foo
       a::Int
       b::Int
    end

    struct Goo
       a::Int
       b::Vector{Float64}
    end
end
```

naively this will need to generate

```julia
struct Demo
    tag::DemoVariantType
    f1::Int
    f2::Int
    f3::Vector{Float64}
end
```

and then for `Goo`, it will use `f1` and `f3`, which requires `f2` to be `uninitialized`, but `new` do not support doing this. 

And this is not true if `Goo.b` is a parametric type instead of `Vector{Float64}`, when it's `Int`, we don't need to generate `f3`, but this is not possible with the current package level support because we cannot generate a struct based on type information.

**Reply by ToucheSir** [reply]
AIUI Unityper.jl has basically the same problems with `getfield` and parametric types. I'd love to see generated structs become a thing too.

**Reply by Roger-luo** [reply]
Julia generates struct in some sense for parametric type - it's just GADT requires a different memory layout based on the type variable value. Unityper does not support namespace it is a less featured ADT for a very specific use case thus it does not have the issue as above.

**Reply by ToucheSir** [reply]
Oh, I think we're 100% in agreement. I meant to point out that Unityper also doesn't support `getfield` because it generates `getproperty`, and because we can't synthesize GADTs from parametric variants it also doesn't support parametric field types.

**Reply by Roger-luo** [reply]
Yeah, I'm just saying Unityper is not generating `getproperty` because it doesn't aim to support namespace at all...

https://github.com/search?q=repo%3AYingboMa%2FUnityper.jl%20getproperty&type=code

which means it support `getfield`?

**Reply by ToucheSir** [reply]
It does generate `getproperty`, see https://github.com/YingboMa/Unityper.jl/blob/ff11ea174eee1058d631e444aaf1a2508c23a2df/src/compactify.jl#L182 and https://github.com/YingboMa/Unityper.jl/blob/ff11ea174eee1058d631e444aaf1a2508c23a2df/src/compactify.jl#L290. I confirmed locally by running `@macroexpand`. Field names in the compacted container struct are mangled, so I presume it does not support `getfield`.

**Reply by Roger-luo** [reply]
No, I didn't mean generating `getproperty(::MyADT, name::Symbol)`, this has no problem and is currently supported well. It is fine to not have `getfield` for an object, because `getproperty` is designed for use case. 

The problem is when you want to support the namespace. I meant generating `getproperty(::Type{MyADT}, name::Symbol)` to support the namespace, which means you are overloading `getproperty` on `DataType`, and Julia `Core` and `Base` are not expecting that, e.g in Expronicon you can have

```julia
julia> using Expronicon.ADT: @adt

julia> @adt D begin
           D1
           D2(::Int)
       end

julia> D1
ERROR: UndefVarError: D1 not defined

julia> D.D1
D.D1
```

this is not about the field, there is nothing needed from the compiler to support the field names, even for GADT currently, you can still implement it with a much larger structure, it's just without type information there is no way to optimize the memory layout. 

See also https://github.com/Roger-luo/Expronicon.jl/issues/60

**Reply by Roger-luo** [reply]
and TL;DR, I have to make `(:name, :super, :parameters, :types, :instance, :layout, :size, :hash, :flags)` as reserved variant names to prevent things not working in `Base` and `Core` which works but not elegant.

**Reply by ToucheSir** [reply]
Oh I see, you're talking about being able to access variants using the ADT name as a namespace. That's pretty low on the list of concerns for me, but I can see how important it would be for a library like Expronicon.

**Reply by Roger-luo** [reply]
I thought so at the beginning too, we don't have it first and just export all the variants, but in practice, we find this gets inconvenient quickly because some variants of the ADT share the same name naturally, and we have to think hard about an alternative name which is the hardest programming problem or wrap every ADT in their own module - basically the same problem with `Base.@enum` and what `EnumX` solves

**Reply by vtjnash** [reply]
Also reserved names include a, b, var, and body. Though the list changes quite regularly, and you can mess up performance for other people quite badly by doing this, so it is very highly discouraged.

**Reply by Roger-luo** [reply]
> Also reserved names include a, b, var, and body. Though the list changes quite regularly, 

I could also just forward non-variant names to `getfield` it would result in a less informative error message but probably not that bad.

> you can mess up performance for other people quite badly by doing this, so it is very highly discouraged.

could you elaborate why overloading `getproperty(::Type{MyType}, name::Symbol)` will mess up performance? I thought this will only affect `MyType`, not other types? Or this creates a problem for the optimization?

**Reply by vtjnash** [reply]
It creates an optimization barrier, which can end up hurting anyone else's performance.

**Reply by Roger-luo** [reply]
Ok, thanks for notifying this, I didn't know overloading this signature would affect other people's code even if it's not defined for other types since it seems to me `getproperty` is just a normal function that forwards to `getfield`. 

I probably still don't get what optimization barrier means, but this seems now giving us another reason that we should have a native ADT rather than package-level support.

**Reply by chriselrod** [reply]
This `getproperty` overload breaks my code doing a `linsolve!`, a fairly small example is here: https://github.com/Roger-luo/Expronicon.jl/issues/77

**Comment by RigobertoPalinskiHJS** [comment]
It's great to see that there are already several package implementations of Algebraic Data Types (ADT) in Julia, and that they are being used for different use cases. However, you raise an interesting point about the limitations of package implementations of ADT, specifically with regards to type instability and memory layout.

Having ADT supported natively in Julia would certainly be beneficial in terms of performance and memory management. However, it is important to carefully consider the design and implementation details of such a feature to ensure that it is consistent with Julia's type system and type inference.

In addition to pattern matching, which is a common feature of ADT in other programming languages, it would also be useful to have some reflection capabilities for built-in ADT. This would allow users to introspect on the structure of the ADT and its fields, which could be useful in certain use cases.

Overall, I think it's a great idea to explore the possibility of adding native support for ADT in Julia, and it's encouraging to see that there are already several package implementations that can serve as a starting point for such a feature. It will be interesting to see how the Julia community approaches this topic and how a native implementation of ADT could be integrated into the language.

**Comment by Tokazama** [comment]
I'm not sure we really need a new `TaggedType<:Type` to accomplish this. As I mentioned in [elsewhere](https://github.com/JuliaLang/julia/issues/47735#issuecomment-1474130350), we mostly just need ways to optimized direct user intervention in inferring unions. We already have union tags but we have no way of directly interacting with them and have to completely load data before we know the type at the user level.

[This LLVM guide](https://llvm.org/docs/HowToSetUpLLVMStyleRTTI.html) is a good read if you want to think about how this is done. Again, we already kind of do this, but haven't settled on a single way of making this happen internally and efficiently.

**Reply by Roger-luo** [reply]
> We already have union tags but we have no way of directly interacting with them and have to completely load data before we know the type at the user level.

Do you mean we are already doing this for `Union`? Could you elaborate on what a potential example would be if what you described in implemented?

> [This LLVM guide](https://llvm.org/docs/HowToSetUpLLVMStyleRTTI.html) is a good read if you want to think about how this is done.

Yeah, I think that looks similar to how I supported ADT in Expronicon, with some limitations due to `new`, `getproperty` etc. as mentioned above.

**Reply by Roger-luo** [reply]
> We already have union tags but we have no way of directly interacting with them and have to completely load data before we know the type at the user level.

I'm not sure I get what you mean here, but do you mean this is just a syntax issue but the compiler actually supports it?

**Comment by Tokazama** [comment]
> Do you mean we are already doing this for Union? Could you elaborate on what a potential example would be if what you described in implemented?

Yes. Rust's Enum type is just a named union instance that uses tag bits, almost exactly how that LLVM guide does it just under the hood.

Now our unions aren't exactly that because as soon as it's not a bits type within a certain size, then we just do a pointer to a boxed value. It may be worth exploring if it's worth having a specific type that adjusts this storage model a bit, taking advantage of quick runtime inference on the bits type.

With the exception of generating optimized methods to access those bit tags, I don't think it's a compiler issue as much at this point. You can jerry rig this with mutable types but immutable types you really need to do this in C first with how stuff is designed right now.

So if someone is really interested in making this happen I'd start with making an ergonomic named and tagged Union type work first. We can do compiler optimizations in places like the "compiler/tfuncs.jl" once we know what works and won't take too long in JIT compilation.

**Reply by Tokazama** [reply]
I'd also bet there's some low hanging fruit for optimizations once these tools are in place. For example, `Union` isn't a great layout for runtime look up but makes sense for building a type system. If we know our ADT doesn't ever change the number of members a random access list for many member might make sense.

**Reply by Roger-luo** [reply]
so IIUC, you are saying a few things to do

- a surface syntax indicating the data model
- a corresponding data model implemented in C that is similar to Union but does not box things

and the compiler itself probably doesn't need many tweaks to get it work without any more complicated optimization. Is that right?

> For example, Union isn't a great layout for runtime look up but makes sense for building a type system.

I assume if we are going in the rust enum direction, then this new data type will only have a known number of variants and the variants will not need to be multiple dispatched, which will basically allow us to look up using if-else or flattened control flow based on how many variants are there, is this what you meant?

**Reply by Tokazama** [reply]
> a corresponding data model implemented in C that is similar to Union but does not box things

That's more of something I think is worth exploring than a strong directive. 

There's going to be pointers in there eventually. For example this would need a pointer for the last variant no matter what.

```julia
@adt MyType begin
    None
    One(::UInt8)
    Many(Vector{UInt})
end
```

It's more a question of how much that extra indirection is worth avoiding and how to do it. I think we are usually turning the above example into a boxed pointer, but if we always have a tagged bit it may be worth the extra work changing the above from

```C
typdef struct {
    jl_taggedvalue_t *ptr
} jl_mytype_t
```

to 

```C
typdef struct {
    uintptr_t *ptr
    uint8_t tag
} jl_mytype_t
```

Then the first two variants would actually be very quick to load and the last one would load the value from heap. I assume this would snowball into figuring out how to get the GC to recognize the last variant.

But all that is premature optimization and could be secondary to getting a very basic working proof of concept working where you build a robust interface around tagged bits using the existing memory model.

> which will basically allow us to look up using if-else or flattened control flow based on how many variants are there, is this what you meant?

Again this is more of a bunch of random thoughts I've had and less of a directive.

I think we bail on anything greater then 3 member unions right now in inference, so accessing this through custom if-else branches may not be terribly useful to the compiler. I personally think the most impressive part about this concept is that it gives users an ergonomic way of limiting the compiler to a couple types. Usually these big if-else chains converge on a common type anyway, so if it could translate to an if-else chain on the tagged bit you could potentially avoid hitting any tricky type inference and necessary loading.

For example, in the following example you could potentially avoid ever having to figure out what the type of `val` is at the compiler level or allocations to load `val`.

```julia
function only(x::MyType)
    val = x.value
    if val isa None
        return false
    elseif val isa One
        return true
    else  # val isa Many
        return false
    end
end
```

**Reply by Roger-luo** [reply]
> But all that is premature optimization and could be secondary to getting a very basic working proof of concept working where you build a robust interface around tagged bits using the existing memory model.

If you are talking about directly emulating ADT using what is supported in Julia currently, that's exactly what I have built in Expronicon. I'm not sure if you have looked at the package, but it supports exactly the same syntax you mentioned above and the data layout you mentioned. It even has pattern match support. I'm not sure how robust the API would that be, but this is why I open this post to discuss if we can have a native solution instead of a package solution on the current data model... 

there still are limitations, like the `new` cannot partially initialize selected fields. With some hack, you can indeed emulate ADT using the current data model, make type inference think it is type stable, and dispatch with pattern match via MLStyle to decide how you dispatch them. But things like `new`, `getproperty`, and proper namespaces are still limited, like mentioned above overloading `getproperty` to provide a namespace currently may mess up optimization, I think if that's the case it is worth being supported natively by the compiler rather than package support. 

In my mind, if we want to have it in the language, as a fundamental data structure we should support a syntax rather than just macros (might be breaking if we use `enum` as a keyword?), but having a macro could be a 1.x-compatible solution, and a few things we need to change at least in terms of emulating the ADT using current data model is 

- letting `new` support keyword arguments to partially initialize selected fields only
- let compiler special case `getproperty` for this ADT type to support namespace

the pattern match and other things can be supported via package in a less optimal way. Then like you said people might be able to explore more optimization opportunities with this.

**Reply by Tokazama** [reply]
By proof of concept I just mean using tagged bits as an intermediary to pattern matching or other functions ons the enclosed data. I'm still not convinced this should be an entirely new type or that we should adopt the common syntax I've seen out there for handling and instantiating these types.

The bare minimum requirement is tagged bits and we already have some rudimentary support for that. So it wouldn't be controversial to start by adding additional ways of interacting with this. Then we can let things progress from there

**Reply by Roger-luo** [reply]
> The bare minimum requirement is tagged bits and we already have some rudimentary support for that. So it wouldn't be controversial to start by adding additional ways of interacting with this.

Like a low-level API accessing this field?

> I'm still not convinced this should be an entirely new type or that we should adopt the common syntax I've seen out there for handling and instantiating these types.

If there is no new syntax, how do we know the user wants to create such a type? or do you mean this should be an implicit optimization for long `Union`?

It's still quite unclear to me what you are trying to propose tbh...

**Reply by Tokazama** [reply]
This should be a fairly uncontroversial addition that would get the bit tag without loading data.

```C
JL_DLLEXPORT jl_value_t *jl_get_nth_field_tag(jl_value_t *v, size_t i)
{
    jl_datatype_t *st = (jl_datatype_t*)jl_typeof(v);
    if (i >= jl_datatype_nfields(st))
        jl_bounds_error_int(v, i + 1);
    jl_value_t *ty = jl_field_type_concrete(st, i);
    unsigned sel;
    if (jl_is_uniontype(ty)) {
        size_t offs = jl_field_offset(st, i);
        if (jl_field_isptr(st, i)) {
            jl_value_t *tyi = jl_typeof(jl_atomic_load_relaxed((_Atomic(jl_value_t*)*)((char*)v + offs)));
            jl_find_union_component(ty, tyi, &sel);
        }
        if (jl_is_uniontype(ty)) {
            assert(!jl_field_isatomic(st, i));
            size_t fsz = jl_field_size(st, i);
            sel = ((uint8_t*)v)[offs + fsz - 1];
        }
    }
    else {
        sel = 0;
    }
    return jl_box_uint8((uint8_t)sel);
}
```

You'd also need to add a version to "builtins.c" and a more optimized version for in "codegen.cpp" and then work on some stuff in "base/compiler.jl" to propagate constants for bounds checking fields, avoiding errors, etc..

It's not a huge leap from there to imitating Rust's enum pattern matching. I've thought about having something like `fieldisa(obj, field, expected_type)` or if you have a special type with a single field you could have `iskind(obj, expected_type)`.

This would change the Rust match pattern from

```Rust
fn only(x: MyType) ->  u8 {
    match x {
        MyType::None => 0,
        MyType::One => 1,
        MyType::Many => 0,
    }
}
```

to

```julia
function only(x::MyType)
    iskind(x, None) && return false
    iskind(x, One) && return true
    iskind(x, Many) && return false
end
```

This doesn't completely settle things because we'd want to eventually provide some sort of extra abilities to this syntax pattern that makes the syntax `fieldisa(obj, field, T) && getfield(obj, field)` know that `T` is returned and avoid unnecessary look ups in the union type. However, it does provide the ground work for a Julia version of ADTs/Enums.

**Reply by vtjnash** [reply]
That seems to be the existing implementation of `typeof` you copied, so I'd recommend to just use `typeof`.

**Reply by Tokazama** [reply]
> That seems to be the existing implementation of `typeof` you copied, so I'd recommend to just use `typeof`.

I'm not sure what you mean by this since it's not at all like the implementation of `jl_typeof`. It takes `jl_get_nth_field` and returns the type tag known at run time.

**Reply by vtjnash** [reply]
> returns the type tag known at run time 

What do you think `typeof` is?

**Reply by Tokazama** [reply]
I mean the union type tag byte that is `UInt8`, not the `jl_taggedvalue` or any of its union components. In the pattern matching example this can be lowered to code that just tests if two `UInt8` are equal because the field, union type, and expected union member type are compile time known and the result of `find_union_component` can be resolved at compile time.

**Reply by vtjnash** [reply]
Have you looked at the lowered code for typeof when used on a value from a field?

**Reply by Tokazama** [reply]
I'm assuming you're talking about `getfield(obj, field) isa expected_type && return getfield(obj, field)` lowering to what I described above and a bit cast. My point has been that we already have the basic components in place to do ADTs and some pattern matching is close to what we do (when fully optimized), we just need to expose some of the internals so that something like Expronicon can build on it to create more sophisticated patterns/methods.

**Comment by MasonProtter** [comment]
Thanks for bringing this up @Roger-luo. After thinking about  this conversation, I've made some updates to [SumTypes.jl](https://github.com/MasonProtter/SumTypes.jl) that I think makes some steps towards the concerns raised here. 


> supporting generic ADT can be problematic without support from Julia's type inference, taking the Option<T> type (similar cases like Result etc. are similar) from rust as an example

SumTypes.jl is able to handle this pretty well now, fully generically and without type instability. 

> I'm not if this is the only solution, but to make it static, the ADT requires a special fixed memory layout, similar to a C union, that is in principle the maximum size of the variant it contains. Currently Expronicon and Unityper implements this as a special structure, which has a size larger than theoretical because there is no access to control the memory layout from Julia (and we probably shouldn't do it...)

My structs are a bit bigger than Expronicon's / Unityper's structs because I can't compactify at macroexpansion time, but IMO they're still quite small. It'd be great to go smaller though. I think the answer to this would be `@generated struct`s.

> The problem is when you want to support the namespace. I meant generating `getproperty(::Type{MyADT}, name::Symbol)` to support the namespace, which means you are overloading `getproperty` on `DataType`, and Julia `Core` and `Base` are not expecting that, e.g in Expronicon you can have
> 
> ```julia
> julia> using Expronicon.ADT: @adt
> 
> julia> @adt D begin
>            D1
>            D2(::Int)
>        end
> 
> julia> D1
> ERROR: UndefVarError: D1 not defined
> 
> julia> D.D1
> D.D1
> ```
and 
> Also reserved names include a, b, var, and body. Though the list changes quite regularly, and you can mess up performance for other people quite badly by doing this, so it is very highly discouraged.

I've sidestepped this by overloading `Base.adjoint` on sum types:
```julia
julia> @sum_type Foo{T} :hidden begin
           A
           B{T}(::T)
       end

julia> A
ERROR: UndefVarError: A not defined

julia> B
ERROR: UndefVarError: B not defined

julia> Foo{Int}'.A
A::Foo{Int64}

julia> Foo'.B(1)
B(1)::Foo{Int64}

julia> let (; B) = Foo'
           B(1)
       end
B(1)::Foo{Int64}
```

**Reply by Roger-luo** [reply]
I'm actually still confused tbh why getproperty is so special? If you can overload adjoint, why we cannot overload getproperty? 

I'll take a look at your solution to see how it supports this, thanks for sharing! We are thinking about a rewrite a MLStyle and hopefully we can have something first-class in MLStyle with all these mentioned above.

**Reply by Tokazama** [reply]
I think it's a pretty clever solution but as I've been thinking about this issue more it seems like a good example of how the current tooling is falling short of clean and intuitive usage. We might need a new type but I'm not quite sure what it would look like.

**Reply by MasonProtter** [reply]
> I'm actually still confused tbh why getproperty is so special? If you can overload adjoint, why we cannot overload getproperty?

So far as I understand 
1) Methods for `getproperty(::Type{T}, ::Symbol)` already exist, unlike `adjoint`.
2) The compiler itself relies on specific behaviour and implementation of those methods, unlike `adjoint`. 

> I'll take a look at your solution to see how it supports this, thanks for sharing! We are thinking about a rewrite a MLStyle and hopefully we can have something first-class in MLStyle with all these mentioned above.

Yes, I would love to see a rewrite of MLStyle! Hopefully something that's more maintainable, I could never figure out what was going on inside it. If you have questions about how SumTypes works, let me know, I can try and write up the flow of stuff.

**Reply by MasonProtter** [reply]
> I think it's a pretty clever solution but as I've been thinking about this issue more it seems like a good example of how the current tooling is falling short of clean and intuitive usage. We might need a new type but I'm not quite sure what it would look like.

Yes, I think there's a few reasons we should still want something like this natively supported in Base. Though, that said, I think if we got https://github.com/JuliaLang/julia/issues/49187 then I think almost everything we want could be done in a package.

**Reply by Tokazama** [reply]
I don't have a fully hashed out proposal but I'd like to see this approached piecewise, instead of just a raw implementation of an algebraic data type. Just off the top of my head, I thing we'd need something like `ClosedUnion` (where it's basically a union that lays everything out linearly and always includes a byte tag to avoid loading boxed pointers), `NamedUnion` to appropriately support dot operator access, and then you can easily create a algebraic data types that outside of Base that function the way described here.

**Reply by MasonProtter** [reply]
If we had `NamedUnion` what would be left to do in a package? That *is* a sum type (i.e. a [disjoint union](https://en.wikipedia.org/wiki/Disjoint_union))

**Reply by Tokazama** [reply]
If people wanted to use the syntax where the declaration of the ADT/sum type also declares new types. I guess you could implement that in a separate package too, but I just assumed that since we have `NamedTuple` it would be natural to include it here.

**Comment by ameligrana** [comment]
I found now this discussion and I'd just like to mention that I built a new package for all of this: https://github.com/JuliaDynamics/DynamicSumTypes.jl. What distinguish this package from others is that it avoids pattern-matching and also if-else statements (mostly). To define a function differently for each (possibly parametrized) variant, it implements a "fake dispatch" mechanism, I think operating this way is a bit more idiomatic and intuitive. It can be also faster than all the other approaches in some cases, because the default option is to merge fields in a unique struct, while if memory efficiency is required it can use a sum type instead. I managed to integrate it in a few packages easily enough because types implemented in this way behave mostly like normal structs.

**Reply by ameligrana** [reply]
I have pushed some changes to DynamicSumTypes.jl because I found a very very simple methodology (it was sad to remove 1000 locs of code I hardly crafted I have to say) which is as fast as what I have previously implemented (or faster) but much more idiomatic. Now users can really dispatch on variants without any loss of performance, and the variants are just normal structs which are created outside of the macro. I think that now the package is really good, memory occupied is much less than in `Union` with immutable structs, in the benchmark in the ReadMe they occupy 1/3 of the memory of `Union` types and these types are even more performant than SumTypes.jl ones in my benchmarks. So I think that maybe we don't really need any `Base` version of sum types? I also found that in Julia 1.11 the methodology works better, I guess it is because of the great Memory PR
